File: DynamoDB-SDK-Examples/python/WorkingWithItems/transact_write.py
from __future__ import print_function # Python 2/3 compatibility
import boto3, json
from botocore.exceptions import ClientError

client = boto3.client('dynamodb', region_name='us-west-2')

try:
    response = client.transact_write_items(
        TransactItems=[
            {
                'Put': {
                    'TableName': 'RetailDatabase',
                    'Item': {
                        'PK': {
                            'S': '123'
                        },
                        'SK': {
                            'S': 'abc'
                        },
                        'User': {
                            'S': 'lhnng'
                        },
                        'Other': {
                            'S': 'Other attribute field'
                        },
                    },
                    'ConditionExpression': 'attribute_not_exists(PK)',
                }
            },
            {
                'Update': {
                    'TableName': 'RetailDatabase',
                    'Key': {
                        'PK': {
                            'S': '789'
                        },
                        'SK': {
                            'S': 'xyz'
                        }
                    },
                    'UpdateExpression': 'SET #o = :val',
                    'ExpressionAttributeValues': {
                        ':val': {
                            'S': 'Hello, updated value'
                        }
                    },
                    'ExpressionAttributeNames': {
                        '#o': 'Other'
                    }
                }
            }
        ]
    )
except ClientError as e:
    print(e)
    print(e.response['Error']['Message'])
else:
    print("Transact Write succeeded")

File: DynamoDB-SDK-Examples/node.js/WorkingWithItems/transact-write.js
/*const AWS = require("aws-sdk");

AWS.config.update({ region: "us-west-2" });

const documentClient = new AWS.DynamoDB.DocumentClient();

async function transactWriteItem() {
  const params = {
    TransactItems: [
      // Update the product status if the condition is met
      {
        Update: {
          TableName: "Products",
          Key: {
            ProductId: "B07J1337PJ42",
          },
          UpdateExpression: "SET ProductStatus = :new_status",
          ConditionExpression: "ProductStatus = :expected_status",
          ExpressionAttributeValues: {
            ":new_status": "SOLD",
            ":expected_status": "IN_STOCK",
          },
          ReturnValuesOnConditionCheckFailure: "ALL_OLD",
        },
      },
      // Create the order if it doesn't already exist
      {
        Put: {
          TableName: "Orders",
          Item: {
            OrderId: "171-3549115-4111337",
            ProductId: "productKey",
            OrderStatus: "CONFIRMED",
            OrderTotal: "100",
          },
          ConditionExpression: "attribute_not_exists(OrderId)",
          ReturnValuesOnConditionCheckFailure: "ALL_OLD",
        },
      },
    ],
    ReturnConsumedCapacity: "TOTAL",
  };

  const response = await documentClient.transactWrite(params).promise();
  return response;
}

// Execute the actions defined as a single all-or-nothing operation
transactWriteItem()
  .then((data) =>
    console.log("TransactWriteItem succeeded:", JSON.stringify(data, null, 2))
  )
  .catch((error) => console.error(JSON.stringify(error, null, 2)));
*/

// This is an example of a simple TransactWriteCommand with the higher level DocumentClient for Amazon DynamoDB

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, TransactWriteCommand } = require("@aws-sdk/lib-dynamodb");

async function transactWriteItems() {
  const client = new DynamoDBClient({ region: "us-west-2" });
  const ddbDocClient = DynamoDBDocumentClient.from(client);
  try {
    return await ddbDocClient.send(
        new TransactWriteCommand({
          TransactItems: [
            // Update the product status if the condition is met
            {
              Update: {
                TableName: "Products",
                Key: {
                  ProductId: "B07J1337PJ42",
                },
                UpdateExpression: "SET ProductStatus = :new_status",
                ConditionExpression: "ProductStatus = :expected_status",
                ExpressionAttributeValues: {
                  ":new_status": "SOLD",
                  ":expected_status": "IN_STOCK",
                },
                ReturnValuesOnConditionCheckFailure: "ALL_OLD",
              },
            },
            // Create the order if it doesn't already exist
            {
              Put: {
                TableName: "Orders",
                Item: {
                  OrderId: "171-3549115-4111337",
                  ProductId: "productKey",
                  OrderStatus: "CONFIRMED",
                  OrderTotal: "100",
                },
                ConditionExpression: "attribute_not_exists(OrderId)",
                ReturnValuesOnConditionCheckFailure: "ALL_OLD",
              },
            },
          ],
          ReturnConsumedCapacity: "TOTAL",
        })
    );
  } catch (err) {
    console.error(err);
  }
}

transactWriteItems()
    .then((data) =>
        console.log("TransactWriteCommand succeeded:", JSON.stringify(data, null, 2)))
    .catch((error) => console.error(JSON.stringify(error, null, 2)));
